packages/data/src/components/use-select/index.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * WordPress dependencies
+ */
+import { createQueue } from '@wordpress/priority-queue';
+import {
+	useLayoutEffect,
+	useEffect,
+	useState,
+	useRef,
+	useMemo,
+	useCallback,
+} from '@wordpress/element';
+import { isShallowEqualObjects } from '@wordpress/is-shallow-equal';
+
+/**
+ * Internal dependencies
+ */
+import useRegistry from '../registry-provider/use-registry';
+import useAsyncMode from '../async-mode-provider/use-async-mode';
+
+/**
+ * Favor useLayoutEffect to ensure the store subscription callback always has
+ * the selector from the latest render. If a store update happens between render
+ * and the effect, this could cause missed/stale updates or inconsistent state.
+ *
+ * Fallback to useEffect for server rendered components because currently React
+ * throws a warning when using useLayoutEffect in that environment.
+ */
+// const useIsomorphicLayoutEffect = typeof window !== 'undefined' ?
+// 	useLayoutEffect : useEffect;
+const useIsomorphicLayoutEffect = typeof window !== 'undefined' ?
+	useLayoutEffect : useEffect;
+
+const renderQueue = createQueue();
+
+export default function useSelect( mapSelect, nextProps ) {
+	const registry = useRegistry();
+	const isAsync = useAsyncMode();
+	const [ mapOutput, setMapOutput ] = useState( mapSelect( registry.select ) );
+	const queueContext = useMemo( () => ( { queue: true } ), [ registry ] );
+
+	const latestMapSelect = useRef( mapSelect );
+	const latestIsAsync = useRef( isAsync );
+	const latestMapOutput = useRef( mapOutput );
+	const isMounted = useRef();
+
+	/**
+	 * Set current values.
+	 */
+	useIsomorphicLayoutEffect( () => {
+		latestMapSelect.current = mapSelect;
+		isMounted.current = true;
+	} );
+
+	/**
+	 * Flush queue if isAsync changes
+	 */
+	useIsomorphicLayoutEffect( () => {
+		latestIsAsync.current = isAsync;
+		renderQueue.flush( queueContext );
+	}, [ isAsync ] );
```

Comment:
Assuming the approach is sound, these could be extracted to internal only named hooks to make it a bit easier to follow in the main exported hook.

Reply:
You can include this in the previous hook:

```js
useIsomorphicLayoutEffect( () => {
	latestMapSelect.current = mapSelect;
	if ( latestIsAsync.current !== isAsync ) {
		latestIsAsync.current = isAsync;
		renderQueue.flush( queueContext );
	}
	isMounted.current = true;
} );
```

Reply:
hmm, ya I guess.  I mostly liked the separation because the flush would only happen if `isAsync` changes (hence the depedency (so removes the need for the conditional check).  So tomahtoes/tomatoes :)

Reply:
Yeah, I was just thinking it could cause some issues with the order of effect clean-up/execution, but I think it's fine.